### PR TITLE
mercurial: 5.9.3 -> 6.0.1, enable check phase

### PR DESCRIPTION
**Pull request by pacien**
###### Motivation for this change

This updates the mercurial package to the latest available version and enables tests.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested, as applicable:
  - [ ] [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - [x] and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - [ ] or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by minijackson — APPROVED** [review]
LGTM

File: pkgs/applications/version-management/mercurial/default.nix [hunk 1]
```diff
@@ -47,6 +62,25 @@ let
     makeFlags = [ "PREFIX=$(out)" ]
       ++ lib.optional rustSupport "PURE=--rust";
 
+    doCheck = stdenv.isLinux;
```
Comment (lukegb) [inline pkgs/applications/version-management/mercurial/default.nix:65]: Please leave a comment as to why you're disabling tests on not-Linux.

Comment (pacien) [reply pkgs/applications/version-management/mercurial/default.nix]: Added and amended!

**Comment by wegank** [comment]
Looks like builds are failing due to test timeouts on aarch64-linux: [log](https://hydra.nixos.org/build/163352241/nixlog/3), [failing jobs](https://hydra.nixos.org/eval/1736641)

**Comment by Izorkin** [comment]
The build throws an error during the checkout phase:
```
running 8 tests
.^[(B.^[(B.^[(B.^[(B.^[(B.^[(B.^[(B.^[(B
test result: ok^[(B. 8 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 2.59s

make[1]: Leaving directory '/tmp/nix-build-mercurial-6.0.1.drv-0/mercurial-6.0.1'
cd tests && python3 run-tests.py --blacklist blacklists/nix --timeout 1800
running 980 tests using 24 parallel processes
sssss.ss..............s..s..s..s.............s.ss.s..s....sss.........ss.........s....s............s......s..s...s...s....ss...........s..s.......ss...............s.........................s...........................$
--- /tmp/nix-build-mercurial-6.0.1.drv-0/mercurial-6.0.1/tests/test-remotefilelog-permissions.t
+++ /tmp/nix-build-mercurial-6.0.1.drv-0/mercurial-6.0.1/tests/test-remotefilelog-permissions.t.err
@@ -43,8 +43,79 @@
   $ umask 002
   $ mkdir $CACHEDIR
   $ hg -q clone --shallow ssh://user@dummy/master shallow2 --config remotefilelog.cachegroup="`id -g -n`"
-  1 files fetched over 1 fetches - (1 misses, 0.00% hit ratio) over * (glob)
+  ** unknown exception encountered, please report by visiting
+  ** https://mercurial-scm.org/wiki/BugTracker
+  ** Python 3.9.9 (main, Nov 15 2021, 18:05:17) [GCC 10.3.0]
+  ** Mercurial Distributed SCM (version 6.0.1)
+  ** Extensions loaded: rebase, remotefilelog, strip
+  Traceback (most recent call last):
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/hgext/remotefilelog/shallowutil.py", line 482, in setstickygroupdir
+      os.chmod(path, 0o2775)
+  PermissionError: [Errno 1] Operation not permitted: b'$TESTTMP/hgcache'
+
+  During handling of the above exception, another exception occurred:
+
+  Traceback (most recent call last):
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/bin/hg", line 61, in <module>
+      dispatch.run()
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/mercurial/dispatch.py", line 144, in run
+      status = dispatch(req)
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/mercurial/dispatch.py", line 250, in dispatch
+      status = _rundispatch(req)
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/mercurial/dispatch.py", line 294, in _rundispatch
+      ret = _runcatch(req) or 0
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/mercurial/dispatch.py", line 470, in _runcatch
+      return _callcatch(ui, _runcatchfunc)
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/mercurial/dispatch.py", line 480, in _callcatch
+      return scmutil.callcatch(ui, func)
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/mercurial/scmutil.py", line 153, in callcatch
+      return func()
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/mercurial/dispatch.py", line 460, in _runcatchfunc
+      return _dispatch(req)
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/mercurial/dispatch.py", line 1273, in _dispatch
+      return runcommand(
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/hgext/remotefilelog/__init__.py", line 580, in runcommand
+      return orig(lui, repo, *args, **kwargs)
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/mercurial/dispatch.py", line 918, in runcommand
+      ret = _runcommand(ui, options, cmd, d)
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/mercurial/dispatch.py", line 1285, in _runcommand
+      return cmdfunc()
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/mercurial/dispatch.py", line 1271, in <lambda>
+      d = lambda: util.checksignature(func)(ui, *args, **strcmdopt)
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/mercurial/util.py", line 1888, in check
+      return func(*args, **kwargs)
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/mercurial/util.py", line 1888, in check
+      return func(*args, **kwargs)
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/hgext/remotefilelog/__init__.py", line 427, in cloneshallow
+      orig(ui, repo, *args, **opts)
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/mercurial/util.py", line 1888, in check
+      return func(*args, **kwargs)
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/mercurial/commands.py", line 1955, in clone
+      r = hg.clone(
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/mercurial/hg.py", line 957, in clone
+      exchange.pull(
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/hgext/remotefilelog/__init__.py", line 357, in pull_shallow
+      setupclient(self.ui, self.unfiltered())
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/hgext/remotefilelog/__init__.py", line 474, in setupclient
+      shallowrepo.wraprepo(repo)
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/hgext/remotefilelog/shallowrepo.py", line 337, in wraprepo
+      makeunionstores(repo)
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/hgext/remotefilelog/shallowrepo.py", line 107, in makeunionstores
+      cachecontent, cachemetadata = makecachestores(repo)
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/hgext/remotefilelog/shallowrepo.py", line 59, in makecachestores
+      cachecontent = contentstore.remotefilelogcontentstore(
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/hgext/remotefilelog/contentstore.py", line 164, in __init__
+      super(remotefilelogcontentstore, self).__init__(*args, **kwargs)
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/hgext/remotefilelog/basestore.py", line 54, in __init__
+      shallowutil.mkstickygroupdir(self.ui, path)
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/hgext/remotefilelog/shallowutil.py", line 509, in mkstickygroupdir
+      setstickygroupdir(path, gid, ui.warn)
+    File "/tmp/nix-build-mercurial-6.0.1.drv-0/hgtests.h6fqo4ry/install/lib/python/hgext/remotefilelog/shallowutil.py", line 485, in setstickygroupdir
+      warn(_(b'unable to chown/chmod on %s: %s\n') % (path, ex))
+  TypeError: %b requires a bytes-like object, or an object that implements __bytes__, not 'PermissionError'
+  [1]
   $ ls -ld $CACHEDIR/11
-  drwxrws* $TESTTMP/hgcache/11 (glob)
+  ls: cannot access '$TESTTMP/hgcache/11': $ENOENT$
+  [2]

 #endif

ERROR: test-remotefilelog-permissions.t output changed
!....s..s............s.....s................................................................s................................................

```

**Comment by pacien** [comment]
@Izorkin On which platform are you seeing this error? The package builds fine on NixOS with sandboxing enabled.

**Comment by Izorkin** [comment]
Checking  x64 platform. Checked in the master branch - the same error. With `sandbox false` not building.

**Comment by pacien** [comment]
The set-uid and set-gid bits might not be supported by your filesystem or are
perhaps disabled. The Mercurial test runner is however supposed to be checking
whether that feature is supported and skip that test if it is not the case…

**Comment by Izorkin** [comment]
I am using zfs filesystem.

**Comment by Izorkin** [comment]
Another error:
```
running 980 tests using 24 parallel processes
ssssssss...............s..s..s..s............s..ss.s..s....sss.........ss.........s....s............s......s.s.....s..s..s.s............s..s.......ss...............s.........................s..........................$
--- /build/mercurial-6.0.1/tests/test-issue6528.t
+++ /build/mercurial-6.0.1/tests/test-issue6528.t.err
@@ -392,7 +392,6 @@
   $ hg up 3
   1 files updated, 0 files merged, 1 files removed, 0 files unresolved
   $ hg st
-  M b.txt
   $ hg debugrevlogindex b.txt
      rev linkrev nodeid       p1           p2
        0       2 05b806ebe5ea 000000000000 000000000000

ERROR: test-issue6528.t output changed
!..............s.....................s...................................................s.....s..s......s......s......s................s........................................s..................s........s...........$
```

**Comment by Izorkin** [comment]
My working wariant:
```
diff --git a/pkgs/applications/version-management/mercurial/default.nix b/pkgs/applications/version-management/mercurial/default.nix
index c3c04da3a69..bab027ffa09 100644
--- a/pkgs/applications/version-management/mercurial/default.nix
+++ b/pkgs/applications/version-management/mercurial/default.nix
@@ -77,6 +77,8 @@ let
       # tests enforcing "/usr/bin/env" shebangs, which are patched for nix
       test-run-tests.t
       test-check-shbang.t
+      test-remotefilelog-permissions.t
+      test-issue6528.t
       EOF

       # extended timeout necessary for tests to pass on the busy CI workers
```
Skip this test - `test-remotefilelog-permissions.t` and `test-issue6528.t`

**Comment by pacien** [comment]
@Izorkin I could not manage to reproduce the two test failures that you have
encountered. Could you report those issues on the Mercurial bug tracker at
https://bz.mercurial-scm.org/? I will then add those to the test skip list
of the package until they are fixed/confirmed to be reliable.

**Comment by Izorkin** [comment]
You have no way of checking tests on a zfs filesystem?
It will be heavy to check on a virtual machine, the tests take a very long time.

**Comment by lukegb** [comment]
I tested building `mercurial` on a zfs root, it built fine. Let me have a look at the hardened profile.

**Comment by Izorkin** [comment]
@lukegb :
```
{ config, lib, pkgs, modulesPath, ... }:
{
  imports = [
    (modulesPath + "/installer/scan/not-detected.nix")
    (modulesPath + "/profiles/hardened.nix")
  ];

```

**Comment by Izorkin** [comment]
I also checked build on another PC. The package build took 4 hours.
Build fails with this error:
```
sss..ss....s...s..........................s.....s.s..s............s..ss.s..s....sss.........ss.........s....s............s......s.s....s...s..s.s............s..s.......ss...............s....
--- /build/mercurial-6.0.1/tests/test-shelve2.t
+++ /build/mercurial-6.0.1/tests/test-shelve2.t#phasebased#abortflag#continueflag.err
@@ -904,33 +904,28 @@
   $ rm .hg/shelved/default.shelve
   $ hg unshelve
   unshelving change 'default'
-  temporarily committing pending changes (restore with 'hg unshelve --abort')
-  rebasing shelved changes
-  merging a
-  warning: conflicts while merging a! (edit, then use 'hg resolve --mark')
-  unresolved conflicts (see 'hg resolve', then 'hg unshelve --continue')
-  [240]
   $ hg abort
-  unshelve of 'default' aborted
+  abort: no unshelve in progress
+  [20]

 Unshelve without .shelve metadata (can happen when upgrading a repository with old shelve)

   $ cat .hg/shelved/default.shelve
-  node=82e0cb9893247d12667017593ce1e5655860f1ac
+  cat: .hg/shelved/default.shelve: $ENOENT$
+  [1]
   $ hg strip --hidden --rev 82e0cb989324 --no-backup
   $ rm .hg/shelved/default.shelve
-  $ hg unshelve
-  unshelving change 'default'
-  temporarily committing pending changes (restore with 'hg unshelve --abort')
-  rebasing shelved changes
-  merging a
-  warning: conflicts while merging a! (edit, then use 'hg resolve --mark')
-  unresolved conflicts (see 'hg resolve', then 'hg unshelve --continue')
-  [240]
+  rm: cannot remove '.hg/shelved/default.shelve': $ENOENT$
+  [1]
+  $ hg unshelve
+  abort: no shelved changes to apply!
+  [20]
   $ cat .hg/shelved/default.shelve
-  node=82e0cb9893247d12667017593ce1e5655860f1ac
+  cat: .hg/shelved/default.shelve: $ENOENT$
+  [1]
   $ hg abort
-  unshelve of 'default' aborted
+  abort: no unshelve in progress
+  [20]

 #endif


ERROR: test-shelve2.t#phasebased#abortflag#continueflag output changed
!....................s....................................s...............s.....................s...................................................s.....s...s.....s......s......s................$
```

**Comment by lukegb** [comment]
Let's move this discussion over to #154001.

**Comment by gbtb** [comment]
After this update of mercurial, TortoiseHG (GUI app for mercurial) is broken because of version mismatch
```
#!python
    ** Mercurial version (6.0.1).  TortoiseHg version (5.9)
    ** Command: 
    ** CWD: /home/artem/Projects/microsoft-di-add-refactoring
    ** Encoding: UTF-8
    ** Extensions loaded: 
    ** Python version: 3.9.9 (main, Nov 15 2021, 18:05:17) [GCC 10.3.0]
    ** System: Linux nixos 5.10.90 #1-NixOS SMP Wed Jan 5 11:40:34 UTC 2022 x86_64
    ** Qt-5.15.3 PyQt-5.15.4 QScintilla-2.13.1
    
    This version of TortoiseHg requires Mercurial version 5.8.n to 5.9.n, but found 6.0.1
```
Probably this two packages should be updated in a lockstep, or TortoiseHG should fixate its mercurial version

**Pull request merged**